File: app/src/main/java/me/ykrank/s1next/view/dialog/PmRequestDialogFragment.java
package me.ykrank.s1next.view.dialog;

import android.os.Bundle;

import javax.inject.Inject;

import io.reactivex.Observable;
import me.ykrank.s1next.App;
import me.ykrank.s1next.R;
import me.ykrank.s1next.data.api.model.Result;
import me.ykrank.s1next.data.api.model.wrapper.AccountResultWrapper;
import me.ykrank.s1next.widget.EditorDiskCache;
import me.ykrank.s1next.widget.track.event.NewPmTrackEvent;

/**
 * A dialog requests to post pm.
 */
public final class PmRequestDialogFragment extends ProgressDialogFragment<AccountResultWrapper> {

    public static final String TAG = PmRequestDialogFragment.class.getName();

    @Inject
    EditorDiskCache editorDiskCache;

    private static final String ARG_TO_UID = "arg_to_uid";
    private static final String ARG_MESSAGE = "message";
    private static final String ARG_CACHE_KEY = "cache_key";

    private static final String STATUS_PM_SUCCESS = "do_success";

    private String cacheKey;

    public static PmRequestDialogFragment newInstance(String toUid, String msg, String cacheKey) {
        App.get().getTrackAgent().post(new NewPmTrackEvent());
        
        PmRequestDialogFragment fragment = new PmRequestDialogFragment();
        Bundle bundle = new Bundle();
        bundle.putString(ARG_TO_UID, toUid);
        bundle.putString(ARG_MESSAGE, msg);
        bundle.putString(ARG_CACHE_KEY, cacheKey);
        fragment.setArguments(bundle);

        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        App.getAppComponent().inject(this);
    }

    @Override
    protected CharSequence getProgressMessage() {
        return getText(R.string.dialog_progress_message_reply);
    }

    @Override
    protected Observable<AccountResultWrapper> getSourceObservable() {
        String toUid = getArguments().getString(ARG_TO_UID);
        String msg = getArguments().getString(ARG_MESSAGE);
        cacheKey = getArguments().getString(ARG_CACHE_KEY);

        return flatMappedWithAuthenticityToken(token ->
                mS1Service.postPm(token, toUid, msg));
    }

    @Override
    protected void onNext(AccountResultWrapper data) {
        Result result = data.getResult();
        if (result.getStatus().equals(STATUS_PM_SUCCESS)) {
            editorDiskCache.remove(cacheKey);
            showShortTextAndFinishCurrentActivity(result.getMessage());
        } else {
            showShortText(result.getMessage());
        }
    }
}

File: app/src/main/java/me/ykrank/s1next/view/dialog/requestdialog/PmRequestDialogFragment.kt
package me.ykrank.s1next.view.dialog.requestdialog

import android.os.Bundle
import io.reactivex.Single
import me.ykrank.s1next.App
import me.ykrank.s1next.R
import me.ykrank.s1next.data.api.model.wrapper.AccountResultWrapper
import me.ykrank.s1next.widget.track.event.NewPmTrackEvent

/**
 * A dialog requests to post pm.
 */
class PmRequestDialogFragment : BaseRequestDialogFragment<AccountResultWrapper>() {

    override fun getProgressMessage(): CharSequence? {
        return getText(R.string.dialog_progress_message_reply)
    }

    override fun getSourceObservable(): Single<AccountResultWrapper> {
        val bundle = arguments!!
        val toUid = bundle.getString(ARG_TO_UID)
        val msg = bundle.getString(ARG_MESSAGE)

        return flatMappedWithAuthenticityToken { token -> mS1Service.postPm(token, toUid, msg) }
    }

    override fun onNext(data: AccountResultWrapper) {
        val result = data.result
        if (result.status == STATUS_PM_SUCCESS) {
            onRequestSuccess(result.message)
        } else {
            onRequestError(result.message)
        }
    }

    companion object {

        val TAG: String = PmRequestDialogFragment::class.java.name

        private val ARG_TO_UID = "arg_to_uid"
        private val ARG_MESSAGE = "message"

        private val STATUS_PM_SUCCESS = "do_success"

        fun newInstance(toUid: String, msg: String): PmRequestDialogFragment {
            App.get().trackAgent.post(NewPmTrackEvent())

            val fragment = PmRequestDialogFragment()
            val bundle = Bundle()
            bundle.putString(ARG_TO_UID, toUid)
            bundle.putString(ARG_MESSAGE, msg)
            fragment.arguments = bundle

            return fragment
        }
    }
}
